src/transformers/models/mplugdocowl/image_processing_mplugdocowl.py

```diff
@@ -0,0 +1,515 @@
+        # xyxy
+        self.image_size = (image_size, image_size) if isinstance(image_size, int) else image_size
+        self.anchors = np.array(
+            [[0, 0, anchor[1] * image_size[1], anchor[0] * image_size[0]]
+             for anchor in anchors]
+        )
+        self.anchor_areas = box_area(self.anchors)
+        self.interpolation = interpolation
+        self.antialias = antialias
+        
+
+    def __call__(self, img, skip_resize=False):
+        # Resize image based on selected anchor
+        #input_image_size = (img.height, img.width)
+        selected_anchor = anchor_rank(self.anchors, self.anchor_areas, (img.size[1], img.size[0]))
+        target_size = self.anchors[selected_anchor][2:].astype(int)  # target width, height
+        if skip_resize:
+            return selected_anchor  # For debug purposes
+        resized_img = img.resize((target_size[0], target_size[1]), resample=self.interpolation)
+        return resized_img, selected_anchor
+
+    def __repr__(self):
+        detail = f"AnchorResize(image_size={self.image_size}, anchor={self.anchors}, interpolation={self.interpolation}, antialias={self.antialias})"
+        return detail
+
+class ShapeAdaptiveCroppingModule:
+    def __init__(self, image_size=(224,224), anchors='grid_9', grid_dict=GRID_DICT, add_global_img: bool = True, 
+                 add_textual_crop_indicator=False):
+        self.image_size = (image_size, image_size) if isinstance(image_size, int) else image_size
+        self.anchors = [tuple(_) for _ in grid_dict[anchors]]
+        self.anchor_max = max(max(_) for _ in self.anchors)
+        self.anchors_areas = [box_area(np.array([[0, 0, w*self.image_size[1], h*self.image_size[0]]])) for w, h in self.anchors]
+        self.add_global_img = add_global_img
+        self.resizer = AnchorResize(image_size=image_size, anchors=self.anchors)
+    def shape_adaptive_crop(self, image: ImageInput):
+
+            image_patches, selected_anchor = self.resizer(image)
+            image_patches = image_patches.convert("RGB")
+
+            h, w = image_patches.size[0],image_patches.size[1]
+            image_patches = np.array(image_patches).reshape(w,h,3)
+        
+            anchor_size = self.anchors[selected_anchor]
+            new_size = (int(anchor_size[1] * image.size[1]), int(anchor_size[0] * image.size[0]))
+            #resized_image = np.array(image.resize(new_size, Image.BICUBIC))
+
+            # Reshape the image
+            num_h, num_w = anchor_size
+            #image_input = np.array(image_patches)
+            image_input = image_patches.reshape(3, num_h, self.image_size[0], num_w, self.image_size[1])
+          
+            # Step 2: Transpose to get the correct order
+            image_input = image_input.transpose(1, 3, 0, 2, 4)
+            image_input = image_input.reshape((-1, self.image_size[0],self.image_size[1],3))
+            image_patches_list = [image_input[i] for i in range(image_input.shape[0])]
+        
```

Comment:
This part I need to abstract from image processing function somehow. But it depends on the selected anchors for cropping

Reply:
I'm not sure I understand why you need to abstract it away: can you write down the logic flow here? It looks like this should work as intended, albeit the instantiation in `MPLUGDocOwlImageProcessor` should pass the necessary parameters to  `ShapeAdaptiveCroppingModule` (that is `image_size`, `anchors`, and so on should be init-dependent in order to be configurable)

Reply:
I mean right now this function only returns images, because this way we can apply the augmentations on the list of images sequentially, (originally, it should also give the patch positions, number of patches)
Also, I am not sure if this should be the output from the image_processing, which is currently {'pixel_values':...}, or we need to add new key something like {'patch_positions':}  
` if do_rescale:
            images = [
                self.rescale(image=image, scale=rescale_factor, input_data_format=input_data_format)
                for image in images
            ]
        #breakpoint()
        if do_shape_adaptive_cropping:
            patch_images = [self.adaptive_crop(image) for image in patch_images]
            images.extend(patch_images[0])
            #breakpoint()
        if is_scaled_image(images[0]) and do_rescale:
            logger.warning_once(
                "It looks like you are trying to rescale already rescaled images. If the input"
                " images have pixel values between 0 and 1, set `do_rescale=False` to avoid rescaling them again."
            )
        if do_normalize:
            images = [
                self.normalize(image=image, mean=image_mean, std=image_std, input_data_format=input_data_format)
                for image in images
            ]
       `

Reply:
to answer the last question: yes, it's totally fine to add new non-standard keys as long as they are handled correctly afterwards! Sounds like a good idea, and then you can use these keys to create your input for the  `.forward()`.
And then, if it is more convenient to access the patch positions and number of patches at this point, can we modify the function so that we have a returned tuple of objects, and then update the images?

Reply:
Do we need to add augmentations for this implementation? Image processors aren't meant to do augmentation, we made this decision as we'd effectively be reinventing the wheel as there's many other great libraries which do this. The image processor is just responsible for taking raw inputs and making them ready for passing to the model for inference

Reply:
I'll let @danaaubakirova answer in detail but I'm 80% sure we are not adding augmentations: it's just that this model crops inputs and organizes them in a certain way before the `.forward()` call to be able to handle arbitrary aspect ratios/resolutions

Reply:
Thanks for the information @amyeroberts! I apologize for the confusion, these are not the actual data augmentations but the transformations needed to preprocess the input, it involves cropping the image in a dynamic way, resizing, normalizing and converting to tensors. More specifically, the goal is to dynamically adjust images to some predefined "anchors" and keep the aspect ratio of the original image.